src/main/java/org/pitest/junit5/JUnit5TestUnitFinder.java

```diff
@@ -97,10 +102,15 @@ private class TestIdentifierListener implements TestExecutionListener {
         private final Class<?> testClass;
         private final TestUnitExecutionListener l;
         private final List<TestIdentifier> identifiers = synchronizedList(new ArrayList<>());
+        private final boolean serializeExecution;
+        private final Map<UniqueId, ReentrantLock> parentCoverageSerializers = new ConcurrentHashMap<>();
+        private final Map<UniqueId, ReentrantLock> coverageSerializers = new ConcurrentHashMap<>();
+        private final ReentrantLock rootCoverageSerializer = new ReentrantLock();
 
         public TestIdentifierListener(Class<?> testClass, TestUnitExecutionListener l) {
             this.testClass = testClass;
             this.l = l;
+            serializeExecution = !(l instanceof NullExecutionListener);

```

Comment:
It's not immediately clear why serialized execution is dependent on the type of the execution listener. Think some comments to explain what is going on are needed.

Reply:
Added